File: PP_c/lista1/zad5.ml
let rec reverse = fun x -> 
  if x == [] 
    then x 
  else (reverse (List.tl x))@[List.hd x];;

let rec palindrome = fun a -> 
  if a == [] then true 
  else if (List.tl a) == [] 
    then true 
  else 
    if (List.hd (reverse (List.tl a)) == List.hd a) 
      then palindrome (List.tl (reverse (List.tl a))) 
    else false ;;

File: PP_c/lista1/zad5.scala
object Klasa {
    def main(args: Array[String]) = {
        val listaLiczb = 1::3::2::1::Nil
        println(palindrome(listaLiczb))
    }
    def palindrome[A] (xs: List[A]): Boolean = {
        if(xs == Nil){
            return true
        }
    
        if(xs.head == reverse(xs).head){
            if(xs.tail == Nil){
                return true
            }
            return palindrome(reverse(xs.tail).tail)
        }
        return false
    }  
    def reverse[A] (xs: List[A]): List[A] = {
        if(xs == Nil){
            return xs
        }
        else{
            return reverse(xs.tail):::List(xs.head)
        }
    }
}